task02/queue.h

```diff
@@ -1,46 +1,177 @@
-
 #pragma once
 
-typedef int data;
+#include <string>
+#include<exception>
 
-class BasicQueue
-{
-public:
+template <typename data>
+struct Node {
+
+	Node():
+		val(0),
+		rhs(NULL),
+		lhs(NULL) {
+	}
+
+	Node(const data& a, Node* l = NULL, Node* r = NULL):
+		val(a),
+		lhs(l),
+		rhs(r) {
+	}
 
-    // Add element to the end
-    virtual void enqueue(data &) = 0;
+	data val;
+	Node<data>* lhs;
+	Node<data>* rhs;
+};
+
+template <typename data>
+class BasicQueue {
+public:
 
-    // Return first element
-    virtual data dequeue() = 0;
+	// Add element to the end
+	virtual void enqueue(const data& a) = 0;
 
-    // Get total count of elements
-    virtual unsigned getLength() = 0;
+	// Return first element
```

Comment:
Try to avoid changing code from examples, especially when you don't need such changes. It decreases readability.